enhancements/api-review/add-ImagePolicy-to-config.openshift.io-v1alpha1.md

```diff
@@ -0,0 +1,585 @@
+        {
+          "type": "sigstoreSigned",
+          "keyData": "dGVzdC1rZXktZGF0YQ==",
+          "rekorPublicKeyData": "dGVzdC1yZWtvci1rZXktZGF0YQ==",
+          "signedIdentity": {
+            "type": "remapIdentity",
+            "prefix": "test-remap-prefix",
+            "signedPrefix": "test-remap-signed-prefix"
+          }
+        }
+      ],
+      "test1.com": [
+        {
+          "type": "sigstoreSigned",
+          "keyData": "dGVzdC1rZXktZGF0YQ==",
+          "rekorPublicKeyData": "dGVzdC1yZWtvci1rZXktZGF0YQ==",
+          "signedIdentity": {
+            "type": "remapIdentity",
+            "prefix": "test-remap-prefix",
+            "signedPrefix": "test-remap-signed-prefix"
+          }
+        }
+      ]
+    },
+    "docker-daemon": {
+      "": [
+        {
+          "type": "insecureAcceptAnything"
+        }
+      ]
+    }
+  }
+}
+```
+
+The merged cluster override policy and namespaced policy in the below `/path/to/policies/testnamespace.json`  will be rolled out. The condensed json string of the file will be updated to the `status.policyJSON` of `mypolicy-2` CR:
+```json
+{
+  "default": [
+    {
+      "type": "insecureAcceptAnything"
+    }
+  ],
+  "transports": {
+    "docker": {
+      "test0.com": [
+        {
+          "type": "sigstoreSigned",
+          "fulcio": {
+            "caData": "dGVzdC1jYS1kYXRhLWRhdGE=",
+            "oidcIssuer": "https://OIDC.example.com",
+            "subjectEmail": "test-user@example.com"
+          },
+          "rekorPublicKeyData": "dGVzdC1yZWtvci1rZXktZGF0YQ==",
+          "signedIdentity": {
+            "type": "remapIdentity",
+            "prefix": "test-remap-prefix",
+            "signedPrefix": "test-remap-signed-prefix"
+          }
+        },
+        {
+          "type": "sigstoreSigned",
+          "keyData": "dGVzdC1rZXktZGF0YQ==",
+          "rekorPublicKeyData": "dGVzdC1yZWtvci1rZXktZGF0YQ==",
+          "signedIdentity": {
+            "type": "remapIdentity",
+            "prefix": "test-remap-prefix",
+            "signedPrefix": "test-remap-signed-prefix"
+          }
+        }
+      ],
+      "test1.com": [
+        {
+          "type": "sigstoreSigned",
+          "keyData": "dGVzdC1rZXktZGF0YQ==",
+          "rekorPublicKeyData": "dGVzdC1yZWtvci1rZXktZGF0YQ==",
+          "signedIdentity": {
+            "type": "remapIdentity",
+            "prefix": "test-remap-prefix",
+            "signedPrefix": "test-remap-signed-prefix"
+          }
+        }
+      ],
+      "test2.com": [
+        {
+          "type": "sigstoreSigned",
+          "keyData": "dGVzdC1rZXktZGF0YQ==",
+          "signedIdentity": {
+            "type": "matchRepoDigestOrExact"
+          }
+        }
+      ]
+    },
+    "docker-daemon": {
+      "": [
+        {
+          "type": "insecureAcceptAnything"
+        }
+      ]
+    }
+  }
+}
+```
+
+#### Notes
+
+ImagePolicy is defined as namespaced CRD. `namespace: openshift-config` is used for cluster policy.
```

Comment:
Why is this better than having a cluster scoped policy? How do i apply policy to the `openshift-config` namespace?

Is there an argument for the policy to only be applied only by an administrator? And therefore would be better as a cluster scoped resource with a namespace selector to allow it to apply to all or multiple namespaces?

Is there a use case where the cluster administrator may want to apply the policy to a subset of namespaces? Is that possible with the current design?

Reply:
We want to support the project/namespaces level policies. We already have some requests for namespaces/project level credential management that we haven't addressed yet.

The current design does not support an imagepolicy for a subset of namespaces. With the CRD coming out as a namespace scoped, the administrator can develop their own tools to create objects for multiple namespaces. Or in the future, we can extend the schema with something like a namespace selector if requested.

Reply:
> Is there an argument for the policy to only be applied only by an administrator? And therefore would be better as a cluster scoped resource with a namespace selector to allow it to apply to all or multiple namespaces?

@mtrmac @saschagrunert What do you think about this?

Reply:
Having it namespaced is a requirement from product management. 

> Is there an argument for the policy to only be applied only by an administrator? And therefore would be better as a cluster scoped resource with a namespace selector to allow it to apply to all or multiple namespaces?

Yes, we evaluated that in the past and came to the conclusion that it would simplify things if it's only cluster scoped. Still, the end user flexibility is prioritized over implementation complexity by our product management.

Reply:
> Having it namespaced is a requirement from product management.

I don't think that's necessarily true. It seems from the context I have that there's a desire to allow namespace level overrides to the global policy, right? The desire from PM isn't that the resource is namespaced, I think that's an implementation detail

So with that, I would expect the cluster wide policy to be a Cluster scoped object, and the namespace policy to be a namespaced object. There is precedent for this, look at RBAC for example with ClusterRoles and Roles. I think that will be a simpler and more intuitive pattern for end users than having to know that a magic namespace configures the entire cluster, where other namespaces just configure for themselves.

This magic namespace concept is adding complexity, which as far as I can tell, has not been solved yet. There's no mention in this document about how I would apply policy to the `openshift-config` namespace only, nor any restrictions on that namespace currently documented. 

Either the `openshift-config` questions need to be answered, or the CRD should become 2 CRDs, a Cluster scoped and a Namespace scoped CRD.

My vote, from a UX perspective, would be 2 CRDs

Reply:
> Is there an argument for the policy to only be applied only by an administrator? And therefore would be better as a cluster scoped resource with a namespace selector to allow it to apply to all or multiple namespaces?

Consider multi-tenant clusters. In that case the tenants might well want to impose specific signature policies for their private namespaces, without the cluster admin caring to impose those policies for everyone.

---

> Is there a use case where the cluster administrator may want to apply the policy to a subset of namespaces? Is that possible with the current design?

I don’t think partially applying a policy _by a single decision-maker_ makes much sense. Unless the cluster is under attack, images in a single on-registry repo (`vendor-registry.example.com/product/server`) are going to come from a single vendor, signed using one policy and process. So, there is, from a global ideal point of view, _one right way_ to verify signatures for images on that on-registry repo (modulo some small security trade-offs, like “do we care about Rekor”).

I can’t see any reason why a cluster admin would say “I know images in `example.com/vendor/product` are signed by $key, and namespaces A,B,C should enforce that, but namespaces D,E,F should not care, and namespaces G,H should require some other key”.

(Contrast that with the multi-tenant case, where there is _still_ conceptually one “right” policy to apply[1], but the tenant’s admin and the clusters’ admin just don’t _care_ the same amount about the policy, so the tenant can configure a per-namespace policy without having to obtain agreement from the cluster admin.)

[1] (This is true as long as we don’t have per-tenant mirrors, allowing every tenant to redirect “the same scope” to a different data source; we don’t have that and I don’t think we should ever have that. Signatures are useful to the extent image names “mean something”, and are not just nicknames that are habitually redirected to completely different destinations.)

Reply:
update the enhancement to include independent ClusterImagePolicy and ImagePolicy CRDs

Reply:
@JoelSpeed PTAL